packages/dds/sequence/src/sharedString.ts

```diff
@@ -260,4 +280,14 @@ export class SharedString extends SharedSegmentSequence<SharedStringSegment> imp
     public getMarkerFromId(id: string): ISegment {
         return this.client.getMarkerFromId(id);
     }
+
+    protected rollback(content: any, localOpMetadata: unknown): void {
+        if (this.previousInsertsCount > 0 && content === this.previousInserts[this.previousInsertsCount - 1]) {
+            const msg = content as IMergeTreeInsertMsg;
+            this.removeRange(msg.pos1, msg.pos1 + (msg.seg as string).length, false);
```

Comment:
this may cause problems. i don't think you can just call this.removeRange. you probably need to call a new method on this.client. The method will need to call into merge tree and dequeue the pending segment group. see ackPendingSegment which is called on incoming op

Reply:
otherwise, the internal queue will get messed up. similarly, you want to ensure the remove doesn't add anything to the queue

Reply:
otherwise, the internal queue will get messed up. similarly, you want to ensure the remove doesn't add anything to the queue

Reply:
I think I fixed it, let me know if it's not right. And if there are any other things I should test.

As for removing the pending segment caused by the removal, I dequeued it, but I could instead suppress creating it in the first place similarly to how sending the op is suppressed in removeRange. Maybe it doesn't matter for the purposes of this prototype.

Reply:
so it's a bit more complicated. when we do local operations we assign special constants to the segment. for insert the segments seq will be UnassignedSequenceNumber, and for remove we set removedSeq to UnassignedSequenceNumber. we have lots of special handling for UnassignedSequenceNumber, as we know anything with that sequence number will eventually get a real sequence number that is higher than any current sequence number. we know this because we know there is an op in flight, and due to total order broadcast we know that inflight ops will be ordered after any currently order op.  we use this property extensively in our merge resolution logic

Reply:
we have other constants we use for local only maintenance changes like: 
refSeq: UniversalSequenceNumber
clientId: LocalClientId
seq: TreeMaintenanceSequenceNumber

on the merge tree methods you can directly control these, but we hide them in client, as client is meant to be the api for the local client, and performs all operations from the perspective of the local client